File: packages/totem-sdk/packages/wots-lease/src/device.ts
/**
 * Device keyspace partitioning — Layer 2.
 *
 * Deterministic address-index range assignment per device slot.
 * Does NOT accept a raw seed; the caller derives the TreeKey in @totemsdk/core.
 *
 * Layout (8 device slots × 8 addresses each = 64 total):
 *   device 0 → addressIndex 0–7
 *   device 1 → addressIndex 8–15
 *   device 2 → addressIndex 16–23
 *   ...
 */

import type { DeviceKeyRange } from './types.js';

const ADDRESSES_PER_DEVICE = 8;
const MAX_DEVICE_SLOTS = 8;

export function allocateDeviceRange(params: {
  deviceSlot: number;
  deviceId: string;
}): DeviceKeyRange {
  if (params.deviceSlot < 0 || params.deviceSlot >= MAX_DEVICE_SLOTS) {
    throw new RangeError(
      `Device slot must be 0–${MAX_DEVICE_SLOTS - 1}, got ${params.deviceSlot}`,
    );
  }
  const startAddressIndex = params.deviceSlot * ADDRESSES_PER_DEVICE;
  const endAddressIndex = startAddressIndex + ADDRESSES_PER_DEVICE - 1;
  return {
    deviceId: params.deviceId,
    startAddressIndex,
    endAddressIndex,
    addressCount: ADDRESSES_PER_DEVICE,
  };
}

export function deviceSlotForAddressIndex(addressIndex: number): number {
  return Math.floor(addressIndex / ADDRESSES_PER_DEVICE);
}

File: packages/totem-sdk/packages/wots-lease/rust/src/device.rs
use crate::types::DeviceKeyRange;

const ADDRESSES_PER_DEVICE: u32 = 8;
const MAX_DEVICE_SLOTS: u32 = 8;

pub fn allocate_device_range(device_slot: u32, device_id: &str) -> Result<DeviceKeyRange, String> {
    if device_slot >= MAX_DEVICE_SLOTS {
        return Err(format!(
            "Device slot must be 0–{}, got {}",
            MAX_DEVICE_SLOTS - 1,
            device_slot
        ));
    }
    let start_address_index = device_slot * ADDRESSES_PER_DEVICE;
    let end_address_index = start_address_index + ADDRESSES_PER_DEVICE - 1;
    Ok(DeviceKeyRange {
        device_id: device_id.to_string(),
        start_address_index,
        end_address_index,
        address_count: ADDRESSES_PER_DEVICE,
    })
}

pub fn device_slot_for_address_index(address_index: u32) -> u32 {
    address_index / ADDRESSES_PER_DEVICE
}
